source/connection-monitoring-and-pooling/tests/README.rst

```diff
@@ -0,0 +1,181 @@
+        resolved[key] = RESOLVE(context, value)
+      return resolved
+  else:
+    return unresolved
+
+
+Examples: if ``context = { foo: 'bar', fizz: 12, buzz: { spam: 'eggs' } }``
+
+- ``RESOLVE(context, { $$ref: [ "foo" ] })`` equals ``"bar"``
+- ``RESOLVE(context, { $$ref: [ "buzz", "spam" ] })`` equals ``"eggs"``
+- ``RESOLVE(context, { foo: { $$ref : ["fizz"] } })`` equals ``{ foo: 12 }``
+
+Spec Test Match Function
+========================
+
+The definition of MATCH or MATCHES in the Spec Test Runner is as follows:
+
+- MATCH takes two values, ``expected`` and ``actual``
+- Notation is "Assert [actual] MATCHES [expected]
+- Assertion passes if ``expected`` is a subset of ``actual``, with the values ``42`` and ``"42"`` acting as placeholders for "any value"
+
+Pseudocode implementation of ``actual`` MATCHES ``expected``:
+
+::
+  
+  If expected is "42" or 42:
+    Assert that actual exists (is not null or undefined)
+  Else:
+    Assert that actual is of the same JSON type as expected
+    If expected is a JSON array:
+      For every idx/value in expected:
+        Assert that actual[idx] MATCHES value
+    Else if expected is a JSON object:
+      For every key/value in expected
+        Assert that actual[key] MATCHES value
+    Else:
+      Assert that expected equals actual
+
+Unit Test Runner:
+=================
+
+For the unit tests, the behavior of a Connection is irrelevant beyond the need to asserting ``connection.id`` and ``connection.generation``. Drivers MAY use a mock connection class for testing the pool behavior in unit tests
+
+For each YAML file with ``style: unit``, for each element in ``tests``:
+
+- Initialize an empty dictionary ``context``
+- Execute each ``operation`` in ``operations``
+
+  - If a ``thread`` is specified, execute in that corresponding thread. Otherwise, execute in the main thread.
+  - If an ``object`` is specified, execute the operation against ``context[object]``
+  - If a ``returnTo`` is specified, set ``context[returnTo]`` to the return value of the operation
+  - If ``args`` are specified:
+
+    - For every ``idx``/``arg`` in ``args``:
+    
+      - ``args[i] = RESOLVE(context, arg)``
+
+    - Pass ``args`` into the operation
+
+- Wait for the main thread to finish executing all of its operations
+- If ``error`` is presented
+
+  - Assert that an actual error ``actualError`` was thrown by the main thread
+  - Assert that ``actualError`` MATCHES ``RESOLVE(context, error)``
+
+- Else: 
+
+  - Assert that no errors were thrown by the main thread
+
+- ``expectedEvents = []``
+- for every ``idx/value`` in ``events``: 
+
+  - ``expectedEvents[idx] = RESOLVE(context, value)``
+
+- calculate ``actualEvents`` as every Connection Event emitted whose ``type`` is not in ``ignore``
+- if ``expectedEvents`` is not empty, then for every (``expectedEvent``, ``i``) in ``expectedEvents``
+
+  - Assert that ``actualEvents[i]`` exists
+  - Assert that ``actualEvents[i]`` MATCHES ``expectedEvent``
```

Comment:
I see that event order is not mandated, was this also intentional for e.g. connection creation event vs wait queue entering event?

Reply:
Event order should be mandated. Every `actualEvent` should line up with a corresponding `expectedEvent`